### PR TITLE
Use 2 GraphQL Executors

### LINKED ISSUE #5501: Support multiple graphql executors
​

Using redis as our pubsub, we should figure out how to pop up multiple executors & have the webserver be able to send a job to one of them.

​

AC

- [ ] a call to graphql executor gets sent to one of the executor instances using a round robin methodology

- [ ] ideally, the message is sent via redis pubsub instead of HTTP

​

Estimate: 8 hours

### LINKED ISSUE #5557: Plan how to scale to 2 GraphQL Executors
We eventually need to scale horizontally. By that I mean our app should support multiple stateful websocket servers that communicate with multiple stateless graphql executors.

I'm not really sure the best way to do that, so I'm learning & documenting what I read here.

### Definitions for our Use Case
- Producer: a websocket server that sends a GQL query to a Consumer.
- Consumer: a graphql executor
- Consumer group: a cluster of consumers

### Evaluated Technologies
- Bull MQ - This is a job queue. It works primarily using redis pub/sub. Pub/Sub means that every consumer receives the message, but only 1 wins. 
- Kakfa - This is probably the most production-ready approach. There are SaaS offerings like NATS.io, but that won't work for our air-gapped solutions. Introducing kafka is possible, but I'd like to limit how many moving parts our app has, if possible.
- Redis streams. These came out in 2018 with Redis v5. They are similar to kafka. Message ordering isn't respected like it is in kafka, but that is fine. it's redis, so it's all in-memory, which is great. It's new-ish, so not as popular as the others. Since we have redis already, this looks like a great answer
- K8s + load balancer. If each consumer is a pod, then we could put a load balancer in front of each cluster & let k8s handle balancing the jobs. this would require minikube or k3s or something similar for development purposes. i think it'd also have to use HTTP vs. just a message bus.

### Learning resources
- https://redis.io/topics/streams-intro
- Redis Microservices for dummies e-book: https://redis.com/blog/how-redis-fits-with-a-microservices-architecture/
- Consumer groups article: https://www.infoworld.com/article/3321938/how-to-use-redis-streams.html

### Implementations
- https://github.com/hoekma/redpop

Kinda crazy that only 1 person has published an abstraction of this pattern!

### PR DESCRIPTION
**Pull request by mattkrick**
TEST
- [x] in your .env set `DEV_NUM_EXECUTORS=2` to spin up multiple executors (this is handled in the pm2 config in production)
- [x] navigate around the app. everything should work as expected. you can `console.log(process.env.SERVER_ID)` to see which graphql executor is handling the job
- [x] when a mutation triggers a publish to the pubsub (e.g. creating a task publishes an event to the TASK channel), every subscriber reuses the same server & dataloader dictionary. for example, if a mutation uses dataloader 5 on gql server 3, every subscriber would use dataloader 5 on server 3 to resolve their payload.

Fix #5557
fix #5501 

### What it does

Before, the webserver and gqlExecutor had a bi-directional pubsub.  

The request pubsub:
1. The webserver published requests to the executor
2. The executor subscribed to those requests

The response pubsub:
3. The executor published responses to the webserver
4. The webserver subscribed to those responses

This PR adds a redis stream to be used in place of the request pubsub.
1. The webserver adds requests to a stream that is capped at 1000 items
2. Each executor instance joins a consumer group that listens to those requests

Notes:
1. The consumer group delegates requests to instances one at a time. Even though multiple requests aren't sent at once today, it's worth making this explicit.
2. The pending entries list (PEL) for the consumer group is disabled, meaning consumers do not need to send an ACK. This means jobs can't become stalled in the event of an error (e.g. timeout while resolving mutation, executor crashes). This is because jobs are NOT idempotent (think GraphQL mutations) and should take no more than a couple seconds. If a job takes longer than that, we alert the user & they can resubmit the job. If we kept a PEL and tried to retry failed jobs ourselves without knowing why a job failed, it's possible that a job half succeeded & we would keep succeeding the first part. For example, imagine if we didn't gracefully catch fetch requests. if the job was to create a task & push it to github & the push to github failed, we'd keep creating new tasks! Best to keep it simple until we build a dead letter queue.
3. The request pubsub is still in place so a subscriber can request that the machine that fulfilled the job of the mutation can fulfill the job of the subscription. 

TODO:
- [x] Support sending a request directly to an instance. By doing this, we can guarantee that all subscription payloads are resolved on the same server as the mutation, thus ensuring the reuse of a dataloader. Without this, we're hitting the database n * m extra times where n is the number of executor instances & m is the number of db hits the mutation response required.
 - [x] Test backpressure (GQL Executor instance is busy & it receives a task, what do?)
 - [ ] 
infra

### CODE REVIEW

**Comment by mattkrick** [comment]
@Dschoordsch ready for your review. If anything is unclear please let me know & I'll do a better job of documenting the flow.

File: packages/gql-executor/gqlExecutor.ts [hunk 1]
```diff
@@ -1,27 +1,58 @@
...
+    ServerChannel.GQL_EXECUTOR_CONSUMER_GROUP,
+    serverChannel
   )
+  console.log(`\n💧💧💧 Ready for GraphQL Execution: ${SERVER_ID} 💧💧💧`)
+
+  for await (const message of incomingStream) {
+    // don't await the call below so this instance can immediately call incomingStream.next()
+    // and be put back in the consumer group, which means it can process more than 1 job at a time
+    // See https://www.loom.com/share/b56812bc561348d0b3d74fe35414499d
+    onMessage('', message)
```
Comment (Dschoordsch) [inline packages/gql-executor/gqlExecutor.ts:54]: I wanted to disagree here, but did some testing and it seems I was wrong and it's fine :+1: 
https://www.loom.com/share/0d8aeea802674a76b0c658e3bcdce7c8

**Review by Dschoordsch — COMMENTED** [review]
I still want to do a bit more testing, but I left you already some comments

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
From the testing I've done it works. Just a couple of nits and then the promisify which seems wrong.
While you're at it, you could also update pm2.dev.config.js with
```
      watch: ['packages/server', 'packages/gql-executor'],
```

File: scripts/gqlServers.js [hunk 2]
```diff
@@ -11,7 +14,20 @@ const gqlServers = async () => {
...
-  require('../dev/gqlExecutor.js')
+
+  // test a cluster in development
+  const executorPath = path.join(__dirname, '../dev/gqlExecutor.js')
+  const fork = promisify(cp.fork)
+  const {SERVER_ID, DEV_NUM_EXECUTORS} = process.env
+  // allow 1 - 8 executors to run in dev
+  const numExecutors = Math.min(Math.max(Number(DEV_NUM_EXECUTORS) || 1, 1), 8)
+  for (let i = 0; i < numExecutors; i++) {
+    const forkServerId = String(Number(SERVER_ID) + i + 1)
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply scripts/gqlServers.js:25]: that's intended, assuming your .env says `SERVER_ID='1'`:
- 1: webserver (web.js)
- 2: gqlExecutor (instance 1)
- 3: gqlExecutor (instance 2) 

i'll add a comment

File: scripts/gqlServers.js [hunk 3]
```diff
@@ -11,7 +14,20 @@ const gqlServers = async () => {
     })
   })
   require('../dev/web.js')
-  require('../dev/gqlExecutor.js')
+
+  // test a cluster in development
+  const executorPath = path.join(__dirname, '../dev/gqlExecutor.js')
+  const fork = promisify(cp.fork)
```
[the remark this answers is not in the document]
Comment (mattkrick) [reply scripts/gqlServers.js:20]: nice catch.
i'm not sure either... i must've been thinking spawn?

File: packages/gql-executor/gqlExecutor.ts (hunk 1 above)
Comment (mattkrick) [reply packages/gql-executor/gqlExecutor.ts:54]: thank you for the sanity check!
i was thinking about keeping a counter of how many jobs are currently processing, but i figured monitoring CPU load will do the same thing without setting an arbitrarily limit

**Pull request merged**